### PR TITLE
[IMPAC-666] Refactor thresholds

### PR DESCRIPTION
**Pull request by xaun**
@cesar-tonnoir please review this refactor for the thresholds. Note this is excluding the threshold tooltips as I am experiencing problems with this and stock charts, but wanted to get at least this refactor finished today.

This refactor is linked to your comments made here: https://github.com/maestrano/impac-angular/pull/393

### CODE REVIEW

File: src/components/widgets-common/chart-threshold/chart-threshold.component.coffee [hunk 1]
```diff
@@ -148,19 +155,25 @@ module.component('chartThreshold', {
...
       widgetHistParams = ctrl.widget.metadata && ctrl.widget.metadata.hist_parameters
       ctrl.disabled = widgetHistParams? && moment(widgetHistParams.to) <= moment.utc().startOf('day')
-      return
+      return !ctrl.disabled
+
+    # Validate and build threshold data from widget kpi templates
+    buildThresholdsFromKpis = ->
+      targets = ctrl.widget.kpis? && ctrl.widget.kpis[0] && ctrl.widget.kpis[0].targets
+      return [] unless ImpacKpisSvc.validateKpiTargets(targets)
+      [{ kpiId: ctrl.widget.kpis[0].id, value: targets.threshold[0].min, name: 'Alert Threshold', color: ctrl.thresholdColor, onClickEvent: onThresholdClick }]
```
Comment (xaun) [inline src/components/widgets-common/chart-threshold/chart-threshold.component.coffee]: @cesar-tonnoir I have left the threshold name static for now as I don't think it makes sense for the widget controller to set this (how would that if / when this is extended to iterate over multiple widget kpis?), and the UI/UX requirements for the name displayed have not been defined as of yet.

**Comment by xaun** [comment]
Firefox is timing out on the build, any ideas? I cannot reproduce this on local. Looking online and not too sure either. I tried adding this with no luck: 
```
# .travis.yml
addons:
 firefox: 'latest'
```

File: src/components/widgets-common/chart-threshold/chart-threshold.component.coffee (hunk 1 above)
Comment (cesar-tonnoir) [reply src/components/widgets-common/chart-threshold/chart-threshold.component.coffee]: ok

**Comment by cesar-tonnoir** [comment]
Regarding the build, this may be because of the changes done on KpisSvc?

**Comment by xaun** [comment]
@cesar-tonnoir please review my latest commit fixing bug discovered during stress-testing. I have stress tested with this commit & working well.

File: src/components/widgets-common/chart-threshold/chart-threshold.component.coffee [hunk 2]
```diff
@@ -123,7 +123,9 @@ module.component('chartThreshold', {
       return unless validateHistParameters()
       Highcharts.addEvent(chart.hc.container, 'click', onChartClick)
       _.each buildThresholdsFromKpis(), (threshold)->
-        ctrl.chart.addThreshold(threshold)
+        thresholdSerie = ctrl.chart.findThreshold(threshold.kpiId)
```
Comment (cesar-tonnoir) [inline src/components/widgets-common/chart-threshold/chart-threshold.component.coffee]: I wouldn't use the kpi id as the key here: couldn't we in the future add thresholds to a chart without them be related to a kpi?
Maybe use the series name instead?

File: src/services/highcharts-factory/highcharts-factory.svc.coffee [hunk 3]
```diff
@@ -102,14 +102,19 @@ angular
         { data: data, showInLegend: false, marker: { enabled: false } },
         thresholdOptions
       )
-      thresholdSerie = @hc.addSeries(threshold)
-      Highcharts.addEvent(thresholdSerie, 'click', (_event)-> thresholdOptions.onClickEvent(thresholdSerie))
-      thresholdSerie
+      @hc.addSeries(threshold)
 
     removeThreshold: (kpiId)->
```
Comment (cesar-tonnoir) [inline src/services/highcharts-factory/highcharts-factory.svc.coffee]: same here: would it work if you were to use the series name instead of the kpi id?

**Pull request merged**